Added mypy to this repo.

nquetschlich opened the pull request:
I still get an error for the `mqt.ddsim` import and do not really understand why it is there, since `mqt.ddsim` is both listed as an additional dependency in the `pre-commit-config.yaml` and also excluded in the `pyproject.toml`. I would have assumed, that either of those two options should resolve the problem:

```
error: Module "mqt" has no attribute "ddsim" 
[attr-defined]
    from mqt import ddsim
```

burgholzer commented:
Hm. I can have a look at that later. Could you, first, merge #15 and rebase this PR on it, so that this PR really only contains changes related to the PR? Makes it easier to check.

A few things as a heads-up. DDSIM is not yet typed. So it definitely needs to be excluded somehow.
The reason why your exclude is not working is because the error already happens at the `mqt` module scope and not at `mqt.ddsim.*`. However, there should be a solution for that.

burgholzer commented:
You will also have to move the `mypy` check to a separate CI workflow similar to how this was handled in the predictor.
This is due to `qiskit-terra` (and it's dependencies) exceeding the memory limits of the pre-commit.ci service.

nquetschlich commented:
> Hm. I can have a look at that later. Could you, first, merge #15 and rebase this PR on it, so that this PR really only contains changes related to the PR? Makes it easier to check.
> 
> A few things as a heads-up. DDSIM is not yet typed. So it definitely needs to be excluded somehow. The reason why your exclude is not working is because the error already happens at the `mqt` module scope and not at `mqt.ddsim.*`. However, there should be a solution for that.

Done. Also added the CI workflow.

burgholzer commented:
> > Hm. I can have a look at that later. Could you, first, merge #15 and rebase this PR on it, so that this PR really only contains changes related to the PR? Makes it easier to check.
> 
> > 
> 
> > A few things as a heads-up. DDSIM is not yet typed. So it definitely needs to be excluded somehow. The reason why your exclude is not working is because the error already happens at the `mqt` module scope and not at `mqt.ddsim.*`. However, there should be a solution for that.
> 
> 
> 
> Done. Also added the CI workflow.

You still need to skip the mypy check in the ci (see the pre-commit Config of the predictor). Otherwise it is run twice.
You can also add back the `mqt.ddsim` package then.

nquetschlich commented:
> > > Hm. I can have a look at that later. Could you, first, merge #15 and rebase this PR on it, so that this PR really only contains changes related to the PR? Makes it easier to check.
> > 
> > 
> > > 
> > 
> > 
> > > A few things as a heads-up. DDSIM is not yet typed. So it definitely needs to be excluded somehow. The reason why your exclude is not working is because the error already happens at the `mqt` module scope and not at `mqt.ddsim.*`. However, there should be a solution for that.
> > 
> > 
> > Done. Also added the CI workflow.
> 
> You still need to skip the mypy check in the ci (see the pre-commit Config of the predictor). Otherwise it is run twice. You can also add back the `mqt.ddsim` package then.

Oh, my bad. I fixed both issues.

burgholzer commented on .pre-commit-config.yaml:
```diff
@@ -79,3 +80,16 @@ repos:
+        args: []
+        additional_dependencies:
+          - types-setuptools
+          - numpy
```
```suggestion
```
Shouldn't be necessary as DDSIM packages qiskit and qiskit packages numpy.

nquetschlich replied:
Shouldn't we be able to remove the `mqt.ddsim` dependency here, since we are ignoring it as configured in the `pyproject.toml`? I tested it locally and it worked with the removed dependency.

burgholzer replied:
While that is the case now, I'd expect `mqt.ddsim` to be typed in the near future. So keeping it in know would mean one change less in the future.
However, I'd be open to remove it here and introduce it back later.

nquetschlich replied:
Makes sense. I keep it.

burgholzer commented on .pre-commit-config.yaml:
```diff
@@ -79,3 +80,16 @@ repos:
+          - types-setuptools
+          - numpy
+          - python_tsp
+          - networkx
```
here we could think about switching to rustworkx (rust port of networkx developed by one of the Qiskit maintainers).
Would allow us to remove the dependency here and (most likely) improve the performance of all graph related stuff.
Nothing for this PR though. Just putting the idea out there.

burgholzer commented on src/mqt/problemsolver/csp.py:
```diff
@@ -1,16 +1,32 @@
-    ):
+        constraints: list[Constraint],
+        quantum_algorithm: str = "Grover",
+    ) -> tuple[int, int, int, int] | bool:
```
Again, nothing for this PR, but to keep in mind for the future: If we really want to expand upon this, dedicated classes or TypedDicts could be employed to make all of this a little more general. 
That holds for the return type here but also for all the tuples of registers down below.
Feel free to just resolve this conversation, but keep the idea in mind.

burgholzer commented on src/mqt/problemsolver/csp.py:
```diff
@@ -202,7 +231,22 @@ def create_oracle(
-    def init_qc(self):
+    def init_qc(
+        self,
+    ) -> tuple[
+        QuantumCircuit,
+        QuantumRegister,
+        QuantumRegister,
+        QuantumRegister,
+        int,
+        int,
+        tuple[
+            tuple[QuantumRegister, QuantumRegister],
+            tuple[QuantumRegister, QuantumRegister],
+            tuple[QuantumRegister, QuantumRegister],
+            tuple[QuantumRegister, QuantumRegister],
+        ],
+    ]:
```
Oh wow, what a monstrosity of a return type 😅
I am fine with it at the moment, but this is also something to be improved in the future.

burgholzer reviewed (approved):
Many thanks for adding mypy here. I just have a couple of really small comments.
Feel free to ignore them and merge as is.

Getting mypy to stop complaining was easier than I thought originally. I just replaced
`from mqt import dddim` with `from mqt.ddsim import <...>`. That way the configuration from the `pyproject.toml` takes effect.
I suspect the original error is due to DDSIM not yet being typed and thus not getting installed as part of the mypy check.
